book/src/dev/rfcs/0005-state-updates.md

```diff
@@ -283,14 +285,16 @@ struct Chain {
 Push a block into a chain as the new tip
 
 1. Update cumulative data members
-    - Add block to end of `self.blocks`
-    - Add hash to `height_by_hash`
-    - for each `transaction` in `block`
-      - add key: `transaction.hash` and value: `(height, tx_index)` to `tx_by_hash`
-    - Add new utxos and remove consumed utxos from `self.utxos`
-    - Add anchors to the appropriate `self.<version>_anchors`
-    - Add nullifiers to the appropriate `self.<version>_nullifiers`
+    - Add the block's hash to `height_by_hash`
     - Add work to `self.partial_cumulative_work`
+    - For each `transaction` in `block`
+      - Add key: `transaction.hash` and value: `(height, tx_index)` to `tx_by_hash`
+      - Add created utxos to `self.created_utxos`
+      - Add spent utxos to `self.spent_utxos`
+      - Add anchors to the appropriate `self.<version>_anchors`
```

Comment:
This is not for each transaction, there will be one for Sprout and one for Sapling for each block, usually determined by last transaction(s) that have JoinSplit or Output descriptions, respectively, with exceptions for genesis and pre-Sapling blocks. These are passed in via the `Request::Commit(Finalized)Block` requests alongside the `Block` itself so maybe `Chain::push()` also needs those pre-computed values.

Reply:
I'll update this in the followup PR that implements the chain_set and the service handling, based on our previous discussion i think it should be "add all the notecommitments in order" to some set